docs/cardano-monads.rst

```diff
@@ -0,0 +1,772 @@
+
+Verdict: the approach is viable but has its costs.
+
+
+Dictionary-Passing Style
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+Similar to previous approach and one was widely used by the networking team.
+
+As we understand ideas behind it:
+
+- Single base monad
+
+  - Similarly to ``ReaderT IO`` approach
+
+- Differentiating implementations all stored in records
+
+  - No type classes and instances like ``instance MyClass (ReaderT Ctx IO)``
+  - Each function accepts a single **explicit** parameter of type ``Ctx m``
+  - This type ``Ctx`` contains all needed methods
+
+Point-by-point rundown:
+
+* Flexibility: LOW. (Same as above).
+
+* Extensibility: GOOD. (Same as above, minus a bit of boilerplate helpers).
+
+* Ease of use: LOW. (Same as above, plus one needs to manually pass a the
+  context everywhere).
+
+* Compile-time performance: SUPERB. (Same as above).
+
+* Run-time performance: MODERATE. (Same as above).
+
+* Run-time configurability: SUPERB. (Same as above)
+
+* Predictability: HIGH. Checking what implementation is used is easy as tracing
+  the provenance of the input context.
+
+Compared to the ``ReaderT``-based approach, explicit dictionary passing trades
+ease of use for simpler types (not a single monad transformer), less boilerplate at
+effect definition site (no need for helpers that ``ask`` and run a method, as
+it's done at use site), and better predictability.
+
+Verdict: preferable to ``ReaderT`` as long as we can put up with reduced
+convenience.
+
+ReaderT over Abstract Base
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+After a discussion, @georgeee and @int-index came up with a hybrid
+approach that combines many elements of other approaches to provide a
+comprehensive solution.
+
+Design goals (in addition to already discussed criteria):
+
+* avoid data types parametrized by an abstract monad ``m``
+  (including explicit method dictionaries),
+  because it makes reasoning about code complicated
+
+* avoid hard-coded ``IO``, ``MonadIO``, ``MonadBaseControl IO``, etc,
+  in order to have good mocking capabilities and separation of concerns
+
+And here are some observations:
+
+* operationally, our monads can be ``ReaderT ctx IO`` for production code
+  and ``ReaderT ctx (ExceptT SomeException ST)`` (or similar) in
+  pure tests. Both of these provide (1) passing context, (2) throwing/catching
+  exceptions, (3) mutable state.
+
+* runtime configuration of effects arises for CLI parameters or YAML configuration
+  (for instance, the slotting method is affected by the '--no-ntp' flag). For this
+  use cases effect context sums must be sufficient.
+
+* presumably, any effect dictionary can be separated into its closure (context)
+  and top-level operations
+
+With these facts established, let us consider the approach from multiple angles:
+how effect definitions would look like, and how effect implementatinos look like.
+
+First, let us have two base monads::
+
+    newtype Prod a = Prod (IO a)
+    newtype Test a = Test (ExceptT SomeException ST a)
```

Comment:
It sounds like a fable. I don't think we will ever be able to emulate everything without `IO`. For example, you need to emulate threads and STM.

Reply:
Hm, good point. I think we might have `TestIO` and `TestPure` to address this, I added them to the doc.